### DISCUSSION TITLE
How to tell if scrolling was initiated by user input?

### DISCUSSION THREAD

**Discussion by szszoke** [inokawa/virtua#413] [Q&A] [answered]
I am using this library to render a list of items. I want to add a "sticky" scroll functionality where if the user scrolls to the bottom then the list is automatically scrolled down to the bottom when new items are added.

If the user scrolls up at any time then the sticky scroll is disabled until they scroll all the way down again.

I was thinking of using the `onRangeChange` callback but my problem is that I cannot tell apart range changes caused by user interaction (i.e. scroll) and other range changes.

Edit:

I tried to use `Virtualizer` with a parent that has `overflow-y: auto`, but the events that I get in `onScroll` have `isTrusted: true` regardless if I used my mouse to scroll or called `scrollToIndex`

**Comment by inokawa** [comment, marked as answer]
The chat example has stick to bottom behavior without distinguishing where the scroll events come from.
https://github.com/inokawa/virtua/blob/main/stories/react/advanced/Chat.stories.tsx
https://inokawa.github.io/virtua/?path=/story/advanced-chat--default

It's difficult to judge a scroll event is from browser or user. Actually, this lib has an internal logic for it but it's not complete and not exposed to userland.

**Reply by szszoke** [reply]
Thanks! I was trying to do something similar myself and this helped me to track down why my logic was failing.

By the way, I noticed that when I call `scrollToIndex` on my list ref for the first time, I get 0 for `viewportSize` in my `onScroll` callback.

I make the call from a `useLayoutEffect` that is triggered when my items array changes.

This only happened if I had enough items in my array to cause scrolling, and I returned to the page where the list was rendered. It ended up being a problem because the logic thought that the user scrolled up and I was storing the auto scroll status in component state instead of a ref.

I'm using component state because I have a switch that shows the status of the auto scrolling.

I just added an if check to return early when `viewportSize` is zero and that fixed this but I am wondering if it is a bug or something that should be handled by the consumer of this library.

**Reply by inokawa** [reply]
viewportSize will be calculated by ResizeObserver so it's 0 until the first measurement.

I agree with that it's inconvenient, but I don't have any plan to improve it for now because it will be difficult without hack.

**Reply by szszoke** [reply]
It is easy enough to work around it and I think it is more important that it is documented.

**Reply by inokawa** [reply]
I updated FAQ section of README.md. Thanks!
